src/Symfony/Component/DependencyInjection/Compiler/AutowiringPass.php

```diff
@@ -0,0 +1,234 @@
+    {
+        if (!($class = $definition->getClass())) {
+            return;
+        }
+
+        foreach ($definition->getTypes() as $type) {
+            $this->definedTypes[$type] = true;
+            $this->typesToId[$type] = $id;
+        }
+
+        if ($reflectionClass = $this->getReflectionClass($id, $definition)) {
+            $this->extractInterfaces($id, $reflectionClass);
+            $this->extractAncestors($id, $reflectionClass);
+        }
+    }
+
+    /**
+     * Extracts the list of all interfaces implemented by a class.
+     *
+     * @param string           $id
+     * @param \ReflectionClass $reflectionClass
+     */
+    private function extractInterfaces($id, \ReflectionClass $reflectionClass)
+    {
+        foreach ($reflectionClass->getInterfaces() as $interfaceName => $reflectionInterface) {
+            $this->set($interfaceName, $id);
+
+            $this->extractInterfaces($id, $reflectionInterface);
+        }
+    }
+
+    /**
+     * Extracts all inherited types of a class.
+     *
+     * @param string           $id
+     * @param \ReflectionClass $reflectionClass
+     */
+    private function extractAncestors($id, \ReflectionClass $reflectionClass)
+    {
+        $this->set($reflectionClass->name, $id);
+
+        if ($reflectionParentClass = $reflectionClass->getParentClass()) {
+            $this->extractAncestors($id, $reflectionParentClass);
+        }
+    }
+
+    /**
+     * Associates if applicable a type and a service id or a class.
+     *
+     * @param string $type
+     * @param string $value A service id or a class name depending of the value of $class
+     */
+    private function set($type, $value)
+    {
+        if (isset($this->definedTypes[$type]) || isset($this->notGuessableTypesToId[$type])) {
+            return;
+        }
+
+        if (isset($this->typesToId[$type])) {
+            if ($this->typesToId[$type] === $value) {
+                return;
+            }
+
+            unset($this->typesToId[$type]);
+            $this->notGuessableTypesToId[$type] = true;
+
+            return;
+        }
+
+        $this->typesToId[$type] = $value;
+    }
+
+    /**
+     * Registers a definition for the type if possible or throws an exception.
+     *
+     * @param \ReflectionClass $typeHint
+     *
+     * @return Reference A reference to the registered definition
+     *
+     * @throws RuntimeException
+     */
+    private function createAutowiredDefinition(\ReflectionClass $typeHint)
+    {
+        if (!$typeHint->isInstantiable()) {
+            throw new RuntimeException(sprintf('Unable to autowire type "%s".', $typeHint->name));
+        }
+
+        $argumentId = sprintf('autowired.%s', $typeHint->name);
+
+        $argumentDefinition = $this->container->register($argumentId, $typeHint->name);
+        $argumentDefinition->setPublic(false);
+
+        $this->definitions = $this->container->getDefinitions();
```

Comment:
instead of maintaining a property ``$this->definitions`` which needs to be updated each time the container is altered to ensure it stays in sync, why not using ``$this->container->getDefinitions()`` directly when needed ?